File: 86_Problem_2.java
// code_report solution
// https://youtu.be/AWLims83GcU

public boolean canVisitAllRooms (List<List<Integer>> rooms) 
{
    Stack<Integer> todo   = new Stack<> (); 
    todo.add (0);
    HashSet<Integer> done = new HashSet<Integer> (); 
    done.add (0);
    
    while (!todo.isEmpty ()) 
    {
        int i = todo.pop ();
        for (int j : rooms.get (i)) {
            if (!done.contains (j)) {
                todo.add (j);
                done.add (j);
            }
        }
    }

    return rooms.size() == done.size();
}

File: 86_Problem_2.py
# code_report Solution
#

def canVisitAllRooms (self, rooms):
   todo = [0]
   done = set (todo)
   while todo:
      i = todo.pop()
      for j in rooms[i]:
            if j not in done:
               todo.append (j)
               done.add (j)

   return len (done) == len (rooms)
